### DISCUSSION TITLE
CodeUnitInsertionException clarification

### DISCUSSION THREAD

**Discussion by ilpescerosso** [NationalSecurityAgency/ghidra#4862] [Q&A] [answered]
Greetings,
Would anyone be able to clarify what the exception in the title means? The full error message was "ghidra.program.model.util.CodeUnitInsertionException: ghidra.program.model.util.CodeUnitInsertionException: Insufficent memory at address 182027d90 (length: 8 bytes)", and it's from a custom ghidra python script if that helps. 

Thanks for reading my question :)

**Comment by astrelsky** [comment, marked as answer]
It means there are not at least 8 bytes of undefined data at address `0x182027d90`. You can use
`clearListing(toaddr(0x182027d90), toAddr(0x182027d90 + (8 - 1)))` before attempting to create the data to fix this.

**Reply by ilpescerosso** [reply]
It seems like the issue persists even after I added a  `addr1 = toAddr(addr)`, 
	`addr2 = toAddr((addr + (8 - 1)))`, 
	`clearListing(addr1, addr2)`  before `createdata(addr, t)`, could it possibly be a problem with my computer?

**Reply by astrelsky** [reply]
> It seems like the issue persists even after I added a `addr1 = toAddr(addr)`, `addr2 = toAddr((addr + (8 - 1)))`, `clearListing(addr1, addr2)` before `createdata(addr, t)`, could it possibly be a problem with my computer?

You must clear enough bytes to fit the size of t. I'm assuming that t here is larger than 8 bytes.

**Reply by ilpescerosso** [reply]
hmmm.. would something like ```clearListing(toAddr(addr), toAddr((addr + (sys.getsizeof(t) - 1)))``` work here? 
Apologies for the basic question, I'm extremely new to ghidra.

**Reply by astrelsky** [reply]
try t.getLength(). For complicated types it may still fail because the size of the type will have changed when added to the program. To avoid this you could try `t.clone(currentProgram.getDataTypeManager()).getLength()`

**Reply by ilpescerosso** [reply]
I currently have:
```	
        len = t.clone(currentProgram.getDataTypeManager()).getLength()
	addr_start = toAddr(addr)
	addr_end = toAddr((addr + (len - 1)))
	clearListing(addr_start, addr_end)
	createData(addr_start, t)
```
where t is "string *" and of length 8, but it's still throwing the exception ```ghidra.program.model.util.CodeUnitInsertionException: ghidra.program.model.util.CodeUnitInsertionException: Insufficent memory at address 182027d90 (length: 8 bytes)```.

**Reply by astrelsky** [reply]
The only other thing I know of that would cause this exception to be thrown is if the memory block itself ends before  `0x182027d90`.

May you log the following;

```python
block = getMemoryBlock(toAddr(addr))
print('block: %s, start: %s, end: %s' % (block.getName(), block.getStart().toString(), block.getEnd().toString()))
```

If it turns out that the block doesn't end before `0x182027d98` then unfortunately I'm all out of ideas.

**Reply by ilpescerosso** [reply]
it turns out you were right, the getMemoryBlock() returned none, thanks for the help :)
